include/zephyr/init.h

```diff
@@ -151,6 +151,39 @@ struct init_entry {
 	__attribute__((__section__(                                           \
 		".z_init_" #level STRINGIFY(prio)"_" STRINGIFY(sub_prio)"_")))
 
+
+/* Designated initializers where added to C in C99. There were added to
+ * C++ 20 years later in a much more restricted form. C99 allows many
+ * variations: out of order, mix of designated and not, overlap,
+ * override,... but C++ allows none of these. See differences detailed
+ * in the P0329R0.pdf C++ proposal. So use '#ifdef __cplusplus' to provide
+ * some separation between these two different languages and to give us
+ * more latitude to support more toolchains without reverting to plain
+ * initializers.
+ */
+#ifdef __cplusplus
+
+/* When using -std=c++2a or higher, g++ (v12.2.0) reject braces for
+ * initializing anonymous unions because it is technically a mix of
+ * designated and not designated initializers which is not allowed in
+ * C++20. Interestingly, the _same_ g++ version does accept braces when
+ * using -std=c++17 or lower!
+ * This Zephyr test compares different -std=... values:
+ *
+ *    ./scripts/twister -v -b -p qemu_cortex_a53 -T tests/lib/cpp/
+ */
+#  define Z_INIT_SYS_INIT_DEV_NULL    .dev = NULL
+
+#else
+
+/* While braces should be optional for anonymous unions, gcc version 4.5
```

Comment:
I think we're all asking what the minimum supported GCC version should be and whether that could be something that doesn't require this hack?